{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Part 1: Libraries installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1579938633.py, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  Cell \u001b[1;32mIn[1], line 2\u001b[1;36m\u001b[0m\n\u001b[1;33m    pip install biopython\u001b[0m\n\u001b[1;37m        ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#Install following libraries\n",
    "!pip install biopython\n",
    "!pip install primer3-py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Part 2: Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import following modules from installed libaries.\n",
    "import Bio\n",
    "from Bio.Seq import Seq\n",
    "import primer3\n",
    "from Bio.SeqUtils.MeltingTemp import Tm_NN\n",
    "#tkinter is required to save file at defined location. it is default library in python.\n",
    "import tkinter as tk\n",
    "from tkinter import ttk\n",
    "from tkinter import filedialog\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Part 3: Primer design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pksingh\\AppData\\Roaming\\Python\\Python38\\site-packages\\Bio\\Seq.py:2880: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results saved to C:/Users/pksingh/OneDrive/01_A-My projects/12_Primer_designing Paper/Data/test_FC_Tools/aprli_24_Class1_2.txt\n"
     ]
    }
   ],
   "source": [
    "#Set a definition for the fast cloning methods: FC1, method to create chimera using two PCRs, 1st for vector, and second for insert. The scheame for FC1 is Vect1-INSERT-Vect2.FC2, method is to insert/delete an small sequence (<100) from a target region. FC2 will generate one pair of primers.\n",
    "#\n",
    "def program_FC1():\n",
    "    #Inputs....\n",
    "    #Ask for insert sequence.\n",
    "    construct_input = input(\"\"\"Paste here your chimera construct <<vector Part1 + Insert + \n",
    "                            Vector part2>>. \n",
    "                            \\n\n",
    "                            Please use only + symbol to separate the sequences. \\n\n",
    "                            The vector part should be more than 40bps and insert 100bp or more\"\"\", )\n",
    "    \n",
    "    # defined split symbol\n",
    "    marker = '+'\n",
    "\n",
    "    #example sequence\n",
    "    #construct_input = \"\"\"ATGAAAGAGAAATTAAACAGACTGCTATTCACGTCGAAAACGAGAATGCGTGTTTTTAGCAAATTATCCAGGTATCTAAGTAATGGTGTTCCAGTAACGTTTGCCCTAGCTGAGCTTTATAAGTTTACGTCTGATGAGGGAAGAAAAAAAGATAATCCGGATGCATTTGCCCTGCAAAGATGGTTAATTGCTGTAAGAAACGGGAAAACATTGGCTGAAGCGATGAGGGGGTGGGTACCTTTTGACGAACTGTCAATAATTTCCGCCGGGGAAATCTCAGGTAATGTTCACCAGGCGTTGGATGATATCATTTATATGAATGATACAAAAAAGAAAGTAAAA   +  ggcgcactggcagggattatttatcctgtagtcctgcttctgacgacatgtctgtatttgcatatatttggaactcaggttgttccggcattttcaggcatcctgcctgtagagaaatggcagggcgcaggcaggactatgtattatcttgctgtattcgttcaggattatcttgtcattacactgctgtcttttatgatggtgatattattaatactggca  +  ACTCTGTCAAGATGGACCGGAAGGTTGCGTCTCTTCTTTGACAGATTTATTCCCTGGTCGATATACAAAACCATTATAGGATGTGGTTTCTTGTTATCACTGGCATCGCTTATTAATGCAGGTATCCCTGTACCGGAAGCCTTACGAATAATAATGAAAACGGCAAGTCCGTGGTATAAGGAAAGATTAGTTGCTATAAGATACAGATTATTAAATGGAGAGAGAAATCTTGGGGAGGCCCTTAATTCATCTGGTTACATTTTCCCATCCAAACAGATGATAATGGATCTCCGTTCTTATGCTGCACTTGGAGGCTTTGATGAAATGCTTAATAAACTTTCAGTTCAATGGCAGGATGACAGTGTTGCTTATATAACGAAACAAATGACT\"\"\"\n",
    "\n",
    "    ## Removed any kind of gap from the input sequence.\n",
    "    cons_str = construct_input.replace(\"\\n\", \"\").replace(' ','').replace('-','')\n",
    "    cons_seq = Seq(cons_str) # to convert into sequence\n",
    "      \n",
    "\n",
    "    # Split the input string from the split symbol \"+\".\n",
    "    sequences = cons_seq.split(marker)\n",
    "\n",
    "    # Confirm if there are exactly three sequences, if not print error.\n",
    "    if len(sequences) != 3:\n",
    "        print(\"Error: Please provide exactly three sequences separated by the specified marker.\")\n",
    "    else:\n",
    "        # Extract and define the individual sequences\n",
    "        vectP1, insert, vectP2 = sequences\n",
    "\n",
    "    # Calculate the \"insert\" sequence length.\n",
    "    len_insert = int(len(insert))\n",
    "    if len_insert <=80:\n",
    "        warning1 = \"Insert length is less than 80bp, you may want use Primer based insertion and select method 2\"\n",
    "    else:\n",
    "        warning1 = None\n",
    "\n",
    "    #Create the final construct by adding sequence as vect_p1 + Insert + vect_p2, also add NNN to avoid any issue during translation into protein sequence in the next step.\n",
    "    \n",
    "    \n",
    "    cons_seq_gap = vectP1+insert.lower()+vectP2 #The constructed sequence, may have any gap!# lower casing insert for user.\n",
    "    \n",
    "    #Remove if there is any gap present.\n",
    "    cons_seq = Seq(cons_seq_gap).replace(\" \", \"\").replace(\"\\n\", \"\")\n",
    "\n",
    "    cons_seq_trans = cons_seq + Seq(\"NNN\") #Added \"NNN\" at the end, it required only for translation.\n",
    "\n",
    "#To convert the sequence into Fasta formate with row length 120 bp.\n",
    "    def format_fasta(sequence, line_length=60):\n",
    "        fasta_string = \">Sequence\\n\"\n",
    "        for i in range(0, len(sequence), line_length):\n",
    "            line = sequence[i:i + line_length]\n",
    "            fasta_string += line + \"\\n\"\n",
    "        return fasta_string\n",
    "\n",
    "   \n",
    "    # Format as FASTA with a line length of 120\n",
    "    fasta_data = format_fasta(cons_seq, line_length=120)\n",
    "\n",
    "    \n",
    "    # Default Tm range\n",
    "    default_min_tm = 55\n",
    "    default_max_tm = 65\n",
    "\n",
    "    # Ask the user if they want to customize Tm values\n",
    "    customize_tm = input(f\"[Default Tm: {default_min_tm}-{default_max_tm}], Do you want to customize the Tm range? (yes/no): \")\n",
    "\n",
    "    if customize_tm.lower() == \"yes\":\n",
    "        # If the answer is yes, ask for desired min and max Tm values\n",
    "        tm_range = input(\"Please provide your 'min-max' Tm range, eg. 60-65\")\n",
    "        min_tm, max_tm = tm_range.split('-')\n",
    "        desired_min_tm = float(min_tm)\n",
    "        desired_max_tm = float(max_tm)\n",
    "    else:\n",
    "        # Use the default values\n",
    "        desired_min_tm = default_min_tm\n",
    "        desired_max_tm = default_max_tm\n",
    "\n",
    "\n",
    "    #Generate translation of the final construct to see, if it is what you expected. This is not niccessory for the fastcloning primer designing but could be useful for user if they want to do protein work.\n",
    "        \n",
    "    #Translate frame 1 for the final construct.\n",
    "    frame1 = cons_seq_trans[0:].translate()\n",
    "\n",
    "    #translate frame 2 for the final construct\n",
    "    frame2 = cons_seq_trans[1:].translate()\n",
    "\n",
    "    #translate frame 3 for the final construct\n",
    "    frame3 = cons_seq_trans[2:].translate()\n",
    "\n",
    "    # Format as FASTA with a line length of 120\n",
    "    fasta_frame1 = format_fasta(frame1, line_length=120)\n",
    "    fasta_frame2 = format_fasta(frame2, line_length=120)\n",
    "    fasta_frame3 = format_fasta(frame3, line_length=120)        \n",
    "\n",
    "    #### Mathmetical parts to do calculation for primer designing.\n",
    "    #First need to find the location (loc1) to start the primers\n",
    "    loc1 = int(len(vectP1)) #Location to start the Reverse Primer for vector amplification.\n",
    "    loc1n = loc1-26 # 26 is a number for the minimum length of primers matching region.\n",
    "    loc2 = loc1+ int(len(insert))\n",
    "\n",
    "    #Lets design the reverse Primer for vector amplifications.\n",
    "    #Vector Revers Primer = pVR.\n",
    "\n",
    "    x = 1 # Want to increase the length by x to achieve required Tm.\n",
    "\n",
    "    pVectR = cons_seq[(loc1n-x):loc1] # Initial reverse primer sequence, this will change later. \n",
    "\n",
    "    # Calculate the initial Tm value\n",
    "    current_tm = round(Tm_NN(cons_seq[(loc1n-x):loc1]))\n",
    "\n",
    "    # Check if the current Tm is outside the desired range\n",
    "    while current_tm < desired_min_tm or current_tm > desired_max_tm:\n",
    "        if current_tm < desired_min_tm:\n",
    "            x += 1  # Increment x to increase Tm\n",
    "        else:\n",
    "            x -= 1  # Decrement x to decrease Tm\n",
    "\n",
    "        # Calculate the new Tm value\n",
    "        current_tm = round(Tm_NN(cons_seq[(loc1n-x):loc1]))\n",
    "\n",
    "    # Now x contains the value that meets the desired Tm range\n",
    "    pVRs = cons_seq[(loc1n-x):loc1]\n",
    "    pVR = pVRs.reverse_complement() #This is the final Reverse primer for Vector.\n",
    "    tmR = round(Tm_NN(pVR)) # Tm for the primer\n",
    "    pVR_len = len(pVR) # Length for the primer.\n",
    "\n",
    "    ############################################\n",
    "    #Forword Primer for vector amplifications\n",
    "    #Vector  Forward Primer = pFR.\n",
    "\n",
    "    y = 1 # Want to increase the length by y to achieve required Tm.\n",
    "    loc3 = loc2+26 #26 is a number for the minimum length of primers matching region.\n",
    "    pVFi = cons_seq[loc2:(loc3+y)] # Initial Forward primer for Vector.\n",
    "    #print(pVFi)\n",
    "    # Calculate the initial Tm value for F primer\n",
    "    current_tmF = round(Tm_NN(cons_seq[loc2:(loc3+y)]))\n",
    "\n",
    "    # Check if the current Tm is outside the desired range\n",
    "    while current_tm < desired_min_tm or current_tmF > desired_max_tm:\n",
    "        if current_tmF < desired_min_tm:\n",
    "            y += 1  # Increment y to increase Tm\n",
    "        else:\n",
    "            y -= 1  # Decrement y to decrease Tm\n",
    "\n",
    "        # Calculate the new Tm value\n",
    "        current_tmF = round(Tm_NN(cons_seq[loc2:(loc3+y)]))\n",
    "\n",
    "    # Now y contains the value that meets the desired Tm range\n",
    "    pVF = cons_seq[loc2:(loc3+y)] # The final Forward primer for Vector. \n",
    "    #print(pVF)\n",
    "    tmf = Tm_NN(cons_seq[loc2:(loc3+y)]) # forward primer Tm\n",
    "    pVF_len = len(pVF) # Forward primer length.\n",
    "    \n",
    "    \n",
    "    ###################################################\n",
    "    ###Design primers for the insert sequence\n",
    "\n",
    "    #Farword Primer for Insert = pIF.\n",
    "\n",
    "    z = 1 #length for matching region sequence from the insert sequence.\n",
    "    ########################################\n",
    "    pIFs = cons_seq[(loc1-16):(loc1+18+z)] # Assumed pIF. It has 16bp from the vectP1 region (from the 3' ends to 16 bp upstream) and from the 3' ends to 22bp + z bp upstream in the insert seq. (loc1+ 22+ z).\n",
    "\n",
    "    # Calculate the initial Tm value for th primerIF\n",
    "    current_tmIF = round(Tm_NN(cons_seq[loc1:(loc1+22+z)]))\n",
    "    # Check if the current Tm is outside the desired range\n",
    "    while current_tmIF < desired_min_tm or current_tmIF > desired_max_tm:\n",
    "        if current_tmIF < desired_min_tm:\n",
    "            z += 1  # Increment z to increase Tm\n",
    "        else:\n",
    "            z -= 1  # Decrement z to decrease Tm\n",
    "        # Ensure z is never less than 0 (minimum matching length of 18 bp)\n",
    "        if z < 0:\n",
    "            z = 0\n",
    "        # Calculate the new Tm value\n",
    "        current_tmIF = round(Tm_NN(cons_seq[loc1:(loc1+22+z)]))\n",
    "\n",
    "\n",
    "    # Now z contains the value that meets the desired Tm range\n",
    "\n",
    "    pIF = cons_seq[(loc1-16):(loc1+22+z)]\n",
    "    tmIf = current_tmIF#round(Tm_NN(cons_seq[loc1:(loc1+z)]),2) #Tm specific for matching region to insert.\n",
    "    tmIf2 = round(Tm_NN(pIF),2) #tm for total length of primer\n",
    "\n",
    "\n",
    "    ##Reverse Primer for Insert = pIR.\n",
    "    w = 1 #length for matching region sequence from the insert sequence.\n",
    "    pIRs = cons_seq[(loc2-22+w):(loc2+16)] # Assumed pIR, It has 22+ w bp from the insert 3' sequenece region + 16bp from VectP2 5' region. The w will get adujusted to meet the Tm.\n",
    "\n",
    "    # Calculate the initial Tm value for the pIR\n",
    "    current_tmIR = round(Tm_NN(cons_seq[(loc2-22+w):loc2]))\n",
    "    # Check if the current Tm is outside the desired range\n",
    "    while current_tm < desired_min_tm or current_tmIR > desired_max_tm:\n",
    "        if current_tmIR < desired_min_tm:\n",
    "            w += 1  # Increment w to increase Tm\n",
    "        else:\n",
    "            w -= 1  # Decrement w to decrease Tm\n",
    "        # Ensure w is never less than 0 (minimum matching length of 18 bp)\n",
    "        if w < 0:\n",
    "            w = 0\n",
    "        # Calculate the new Tm value\n",
    "        current_tmIR = round(Tm_NN(cons_seq[(loc2-22+w):loc2]))\n",
    "\n",
    "    pIRt = cons_seq[(loc2-22+w):(loc2+16)]\n",
    "    pIR = pIRt.reverse_complement() #The final sequence for pIR.\n",
    "    tmIr = current_tmIR # Tm value for the matching sequences from Insert region.\n",
    "    tmIr2 = round(Tm_NN(pIR),2) #tm for total length of primer\n",
    "\n",
    "    # For input data Format as FASTA with a line length of 120\n",
    "    fasta_vectP1 = format_fasta(vectP1, line_length=120)\n",
    "    fasta_insert = format_fasta(insert, line_length=120)\n",
    "    fasta_vectP2 = format_fasta(vectP2, line_length=120)\n",
    "\n",
    "    # Your results\n",
    "    results_FC1 = f\"\"\"Thanks for using FastCloneAssist Primer Tool!\\n\n",
    "    Report:\\nPlease find pairs of Primers to do your Fast_cloning PCR.\\n\n",
    "    Warring: {warning1}\n",
    "    \\n\n",
    "\n",
    "    Vector specific Primer Pair:-\n",
    "    ======================================================\n",
    "\n",
    "        Vect.Primer_F:\n",
    "        5' {pVF} 3'\n",
    "        Primer length: {len(pVF)}\n",
    "        Tm is: {round(tmf, 2)} Degree Celsius\n",
    "\n",
    "        Vect.Primer_R:\n",
    "        5' {pVR} 3'\n",
    "        Primer length: {len(pVR)}\n",
    "        Tm is: {round(tmR, 2)} Degree Celsius\n",
    "\n",
    "    Insert specific Primer Pair:-\n",
    "    ======================================================\n",
    "        \n",
    "        InsertPrimer_F:\n",
    "        5' {pIF} 3'\n",
    "        Primer length: {len(pIF)}\n",
    "        Tm for matching to insert seq is: {round(tmIf, 2)} Degree Celsius\n",
    "        Tm for full-length is: {tmIf2} Degree Celsius\n",
    "\n",
    "        InsertPrimer_R:\n",
    "        5' {pIR} 3'\n",
    "        Primer length: {len(pIR)}\n",
    "        Tm for matching to insert seq is: {round(tmIr, 2)} Degree Celsius\n",
    "        Tm for full-length is: {tmIr2} Degree Celsius\n",
    "    ======================================================\n",
    "\n",
    "    The new construct sequence: The inserted sequece are in lower case for easy viewing.\n",
    "        \\n{fasta_data}\n",
    "\n",
    "    Protein translations of the Final constructs: \n",
    "\n",
    "        Frame 1: \\n {fasta_frame1}\\n\n",
    "        Frame 2: \\n {fasta_frame2}\\n\n",
    "        Frame 3: \\n {fasta_frame3} \\n\n",
    "        -------- For Record Keeping ----------\\n\n",
    "        \n",
    "    Your Input sequences are copied below in the format of- \n",
    "        \\n'Vector_part1+INSERT+Vector_part2':\\n\n",
    "\n",
    "        Vector Part 1:\n",
    "        {fasta_vectP1}\\n\n",
    "\n",
    "        Insert Sequence:\n",
    "        {fasta_insert}\\n\n",
    "\n",
    "        Vector Part 2:\n",
    "        {fasta_vectP2}\n",
    "        \"\"\"\n",
    "\n",
    "    #Print results as display and save in downloads\n",
    "    print(results_FC1)\n",
    "\n",
    "    #To save as Text file in download\n",
    "    from google.colab import files\n",
    "\n",
    "# Generate your report content as a string\n",
    "    report_content1 = results_FC1\n",
    "\n",
    "# Create a text file object\n",
    "    report_file = open(\"FastCloning1_primers.txt\", \"w\")\n",
    "    report_file.write(report_content1)\n",
    "    report_file.close()\n",
    "\n",
    "# Download the file\n",
    "    files.download(\"FastCloning1_primers.txt\")   \n",
    "#########################################\n",
    "#########################################\n",
    "##########################################\n",
    "#Define the FC2 program. This is for Primer based deletion or insertion or both (max insertion sequence 120bp).\n",
    "def program_FC2():\n",
    "    # Inputs, Sequences in organised format and Tm (optional)\n",
    "    # Ask for input, here user will define the deletion region by * symbols as well as if there is insert add * and your insert. see below example. And if there is no deletion only insertion, mark the insertion point by two stars, **, and provide the insert sequence as above.\n",
    "    #construct_inputs = \"\"\"AGATCTCGATCCCGCGAAATTAATACGACTCACTATAGGGGAATTGTGAGCGGATAACAATTCCCCTCTAGAAATAAGGGCGGTGCGTGTGCCGCCTTTGGGG**CGAGCGGGCCGTGTCGGGGGGAATTAATACGACTCACTATAGGGGAATTGTGAGCGGATAACAATTCCCCTCTAGAAATAATTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTATCATCACAGCAGCGGCCTGGTGCCGCGCGGCAGCCATATG *AAAAAAAAAAAAAAAAAAAAAAAAAA\"\"\"\n",
    "    construct_inputs = input(\"\"\"Input sequence, Decorate your sequence with three * symbols. \n",
    "                             \\n\n",
    "                             First two stars to mark deletion sequence boundaries or the insertion location when no deletion needed. The third star at the end of the target sequence. If there is an insert sequence, it will go after the 3rd star.\n",
    "                             \\n \n",
    "                             Example 1: ...ATGCATGC*TTTTT*ATGCATGC...*AAAAA, here TTTTT is the deletion target and AAAAA is the insertion seq. \n",
    "                             \\n\n",
    "                             Example 2: ...ATGC**TGAC...*AAAAA, ** is the insertion location for the AAAAA seq. \n",
    "                             \\n\n",
    "                             Example 3: ...ATGC*AAAA*GTCA...*, here the AAAA is selected for deletion without any insertion\"\"\",\n",
    "                             )\n",
    "    \n",
    "    \n",
    "    ## Remove any gap or line gap from the sequence.\n",
    "    cons_str = construct_inputs.replace(\"\\n\", \"\").replace(\" \",\"\").replace(\"-\",\"\")\n",
    "    #convert string input into a sequence file\n",
    "    cons_seq = Seq(cons_str)\n",
    "    \n",
    "    # Split the input string into parts using the * symbol\n",
    "    parts = cons_seq.split('*')\n",
    "    \n",
    "\n",
    "    # Initialize variables to store the sequences\n",
    "    partP1 = \"\"\n",
    "    deleted_seq = \"\"\n",
    "    partP2 = \"\"\n",
    "    insertP = \"\"\n",
    "\n",
    "    # Check the number of parts, if missing generate a worning to the user.\n",
    "    if len(parts) == 3:\n",
    "        partP1, deleted_seq, partP2 = parts\n",
    "    elif len(parts) == 4:\n",
    "        partP1, deleted_seq, partP2, insertP = parts\n",
    "    else:\n",
    "        print(\"Error: Please provide either 3 or 4 sequences separated by the specified markers.\")\n",
    "\n",
    "    if len(insertP) >= 120:\n",
    "        warning = \"Warning!! The insert is bigger than 120bp\"\n",
    "    else:\n",
    "        warning = None  # No warning if the condition is not met\n",
    "\n",
    "#To convert the sequence into Fasta formate with row length 120 bp.\n",
    "    def format_fasta(sequence, line_length=60):\n",
    "        fasta_string = \">Sequence\\n\"\n",
    "        for i in range(0, len(sequence), line_length):\n",
    "            line = sequence[i:i + line_length]\n",
    "            fasta_string += line + \"\\n\"\n",
    "        return fasta_string\n",
    "\n",
    "    #change part1 and part2 into fasta formate\n",
    "    fasta_partP1 = format_fasta(partP1, line_length = 120)\n",
    "    fasta_partP2 = format_fasta(partP2, line_length = 120)\n",
    "    \n",
    "    ##########################################################\n",
    "    #primer designing: Mathmetical portion\n",
    "\n",
    "    #create the chimera by deleting the region specified (the * marking) and adding the insert sequence there.\n",
    "    geneM_gap = partP1 + insertP.lower() + partP2 #used lower font style to separate the insert, will be helpful to user.\n",
    "    \n",
    "    #Remove any gap further\n",
    "    geneM = Seq(geneM_gap).replace(\"\\n\", \"\").replace(\" \", \"\").replace(\"-\",\"\")\n",
    "\n",
    "    geneM_trans = geneM + Seq(\"NNN\") #Adding \"NNN\" will make translation by the python error free. \n",
    "\n",
    "# Format as FASTA with a line length of 120\n",
    "    fasta_geneM = format_fasta(geneM, line_length=120)\n",
    "\n",
    "    #Translate the modified gene into protein to show, if the insertion is as you wanted.\n",
    "    #translate frame 1 for the final construct.\n",
    "    frame1 = geneM_trans[0:].translate()\n",
    "    #translate frame 2 for the final construct\n",
    "    frame2 = geneM_trans[1:].translate()\n",
    "    #translate frame 3 for the final construct\n",
    "    frame3 = geneM_trans[2:].translate()\n",
    "    #print(geneM_trans, '\\n', frame1, '\\n',frame2,'\\n', frame3)\n",
    "    \n",
    "    #convert translaation into fasta formate\n",
    "    fast_frame1 = format_fasta(frame1, line_length = 120)\n",
    "    fast_frame2 = format_fasta(frame2, line_length = 120)\n",
    "    fast_frame3 = format_fasta(frame3, line_length = 120)\n",
    "\n",
    "\n",
    "\n",
    "    #Input sequences length calculations\n",
    "\n",
    "    partP1_len = int(len(partP1)) # length of part 1, the value is required to do primer start-end calculations.\n",
    "    \n",
    "    p1l = m1 = partP1_len #making short variable.\n",
    "    #find the middle of inserted sequence, it is required to make the primers start & end points.\n",
    "    #calculate the length fo insert\n",
    "    insertP_len = len(insertP) #insert length (il) (insert nucleotide sequence).\n",
    "    \n",
    "    il = insertP_len #made the variable short.\n",
    "    if il % 2 == 0:  #if the division by 2 is not an integer make an integer by adding 1.\n",
    "        il_m = int(il / 2) #to calculate the middle of insert.\n",
    "    else:\n",
    "        il_m = int((il + 1) / 2) #Half of insert seq length: l2\n",
    "    midP = m2 = int(p1l + il_m) # middle position of the insert in the modified sequence, geneM.\n",
    "\n",
    "    oE = m3= int(m2 + (il-il_m)) # oE, other End is the start point of gene after the insert sequence in the geneM.\n",
    "\n",
    "    #print(p1l, il,il_m, midP,oE)\n",
    "\n",
    "    # Default Tm range\n",
    "    default_min_tm = 55\n",
    "    default_max_tm = 65\n",
    "\n",
    "    # Ask the user if they want to customize Tm values\n",
    "    customize_tm = input(f\"Do you want to customize the Tm range? (yes/no) [Default: {default_min_tm}-{default_max_tm}]: \")\n",
    "\n",
    "    if customize_tm.lower() == \"yes\":\n",
    "        # If the user wants to customize, ask for desired min and max Tm values\n",
    "        tm_range = input(\"Please provide your 'min-max' Tm range, eg. 60-65\")\n",
    "        min_tm, max_tm = tm_range.split('-')\n",
    "        desired_min_tm = float(min_tm)\n",
    "        desired_max_tm = float(max_tm)\n",
    "    else:\n",
    "        # Use the default values\n",
    "        desired_min_tm = default_min_tm\n",
    "        desired_max_tm = default_max_tm\n",
    "\n",
    "    ########################\n",
    "    ############################\n",
    "    #Forward primer (p1f) design....\n",
    "\n",
    "    x = 1  # starting with 18bp, but it will change as per Tm needs.\n",
    "    #Forward primer_match region seq, see below remarks.\n",
    "    \n",
    "    p1f = geneM[m3:(m3 +26+(x))]\n",
    "\n",
    "    # Calculate the initial Tm value for F primer\n",
    "    current_tm_p1f = round(Tm_NN(geneM[m3:(m3 +26 + (x))]))\n",
    "    # Check if the current Tm is outside the desired range\n",
    "    while not (desired_min_tm <= current_tm_p1f <= desired_max_tm):\n",
    "        if current_tm_p1f < desired_min_tm:\n",
    "            x += 1  # Increment x to increase Tm\n",
    "        else:\n",
    "            x -= 1  # Decrement x to decrease Tm\n",
    "\n",
    "        # Ensure x is never less than 0 (minimum matching length of 18 bp)\n",
    "        if x < -8:\n",
    "            x = 7\n",
    "\n",
    "        # Calculate the new Tm value\n",
    "        current_tm_p1f = round(Tm_NN(geneM[m3:(m3 +26+(x))]))\n",
    "    p1fs = geneM[m3:(m3 +26+(x))]\n",
    "    #print(p1f)\n",
    "    #print(len(p1f))\n",
    "    #print(p1fs)\n",
    "    #print(len(p1fs))\n",
    "    #print(current_tm_p1f)\n",
    "    #print(x)\n",
    "\n",
    "    #(pmf), Forward primer's region between overhang(p2f) and match(p1f),\n",
    "    pmf = geneM[m2:m3]  #this length will vary according to the insert length.\n",
    "    #Forward primer_overhang (p2f) = geneM[(m2-8):m2]\n",
    "    p2f = geneM[(m2 - 8):m2]\n",
    "\n",
    "    #Now get the complete forward primer by adding the p2f and p1f part.\n",
    "    primerF = p2f + pmf + p1fs\n",
    "\n",
    "    #primer length\n",
    "    pf_l = len(primerF)\n",
    "    primerFtm = round(Tm_NN(primerF), 2) #for full length\n",
    "    #print(primerF)\n",
    "    #print(pf_l,primerFtm)\n",
    "    #print(current_tm_p1f)\n",
    "\n",
    "    #Revers primer designing, same as above, only need to convert as complementry seq at the end.\n",
    "\n",
    "    \n",
    "    y = 1  # staring with 1, but it will change as per Tm needs.\n",
    "\n",
    "    #Reverse primer_match (p1r)\n",
    "    p1r = geneM[(m1 -(26 + y)):m1]\n",
    "\n",
    "    # Calculate the initial Tm value for F primer\n",
    "    current_tm_p1r = round(Tm_NN(geneM[(m1 -(26 + y)):m1]))\n",
    "    # Check if the current Tm is outside the desired range\n",
    "    while not (desired_min_tm <= current_tm_p1r <= desired_max_tm):\n",
    "        if current_tm_p1r < desired_min_tm:\n",
    "            y += 1  # Increment x to increase Tm\n",
    "        else:\n",
    "            y -= 1  # Decrement x to decrease Tm\n",
    "\n",
    "        # Ensure x is never less than 0 (minimum matching length of 18 bp)\n",
    "        if y < -8:\n",
    "            y = 7\n",
    "\n",
    "        # Calculate the new Tm value\n",
    "        current_tm_p1r = round(Tm_NN(geneM[(m1 -(26 + y)):m1]))\n",
    "    p1rs = geneM[(m1 -(26 + y)):m1]\n",
    "    #print(p1r)\n",
    "    #print(len(p1r))\n",
    "    #print(p1rs)\n",
    "    #print(len(p1rs))\n",
    "    #print(current_tm_p1r)\n",
    "    #print(y)\n",
    "\n",
    "    #(pmr), Reverse primer's region between overhang(p2r) and match(p1rs),\n",
    "    pmr = geneM[m1:m2]  #this length will vary according to the insert length.\n",
    "\n",
    "    #Revers primer_overhang (p2r) = geneM[m2:(m2+8)]\n",
    "    p2r = geneM[m2:(m2 + 8)] # 8pb should be matched\n",
    "    #print(p2r)\n",
    "    primerR = (p1rs + pmr + p2r).reverse_complement()\n",
    "    #print(f\"Reverse primer: 5' {primerR} 3'\")\n",
    "    #primer length\n",
    "    pr_l = len(primerR)\n",
    "    #primerRtm = round(Tm_NN(primerR[(pr_l - y):]), 2)\n",
    "    primerR_tm = round(Tm_NN(primerR),2)\n",
    "    #print(primerR_tm)\n",
    "    #print(\"Reverse primer length is: \", (pr_l), \"bp\")\n",
    "    #Generate reports\n",
    "    # Your results\n",
    "    results_FC2 = (\n",
    "        \"\"\"Thanks for using FastCloneAssist Primer Tool!\\n\\n\"\"\"\n",
    "        f\"Report: Your primers for performing Fast Cloning PCR to either delete, insert, or delete and insert at a specified location..\\n\\n\"\n",
    "        f\"Warning: {warning}\\n\\n\"\n",
    "        f\"=====================================================================\\n\\n\"\n",
    "        f\"Forward primer:\\n 5' {primerF} 3'\\n\"\n",
    "        f\"Tm based on total length of the primer is: {primerFtm} Degree Celsius\\n\"\n",
    "        f\"Tm for matching seq is: {current_tm_p1f} Degree Celsius\\n\"\n",
    "        f\"Forward primer length is: {pf_l} bp\\n\\n\"\n",
    "        f\"Reverse primer:\\n 5' {primerR} 3'\\n\"\n",
    "        f\"Tm based on total length of the primer is: {primerR_tm} Degree Celsius\\n\"\n",
    "        f\"Tm for matching seq is: {current_tm_p1r} Degree Celsius\\n\"\n",
    "        f\"Reverse primer length is: {pr_l} bp\\n\"\n",
    "        f\"=====================================================================\\n\\n\"\n",
    "        f\"The modified Chimera gene sequence:\\n{fasta_geneM}\\nlength = {(len(geneM))}bp\\n\\n\"\n",
    "        f\"Final Sequence Translation, Frame 1:\\n{fast_frame1}\\n\\n\"\n",
    "        f\"Final Sequence Translation, Frame 2:\\n{fast_frame2}\\n\\n\"\n",
    "        f\"Final Sequence Translation, Frame 3:\\n{fast_frame3}\\n\\n\"\n",
    "        f\"----------For The Record Keeping-------------\\n\"\n",
    "        f\"Your Input sequence after partition are: \\nSequence Part 1: \\n{fasta_partP1}\\n\\n\"\n",
    "        f\"Sequence Part 2: \\n{fasta_partP2} \\n\\n\"\n",
    "        f\"The chosen deletion region length = {(len(deleted_seq))}bp \\nDeleted sequence: \\n{deleted_seq}\\n\\n\"\n",
    "        f\"The provided insert sequence is:\\n{insertP}\\nlength = {(len(insertP))}bp\\n\\n\"\n",
    "        \n",
    "        \n",
    "    )\n",
    "#Print results as display and save in downloads\n",
    "    print(results_FC2)\n",
    "\n",
    "    #To save as Text file in download\n",
    "    from google.colab import files\n",
    "\n",
    "# Generate your report content as a string\n",
    "    report_content2 = results_FC2\n",
    "\n",
    "# Create a text file object\n",
    "    report_file = open(\"FastCloning2_primers.txt\", \"w\")\n",
    "    report_file.write(report_content2)\n",
    "    report_file.close()\n",
    "\n",
    "# Download the file\n",
    "    files.download(\"FastCloning2_primers.txt\") \n",
    "\n",
    "    \n",
    "#And you can make def for your program now it can be asked as choice,..\n",
    "\n",
    "choice = input(\"\"\"\\nWELCOME to MD:FastCloning_PrimerTool!\\nSelect your Method of Fast cloning.\\nEnter 1, if you want to PCR amplify your Vector and Insert separately, you will get 2 pairs of primers.\\nEnter 2, if you want to do primer based insertion (<120bp) or/and deletion\"\"\",)\n",
    "\n",
    "if choice == \"1\":\n",
    "    program_FC1()\n",
    "elif choice == \"2\":\n",
    "    program_FC2()\n",
    "else:\n",
    "    print(\"Invalid choice. Please select 1 or 2.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
